### PR TITLE
Add syntactic rewriting

### PR DESCRIPTION
**Pull request by jcrist**
Allows for creation of a ``RuleSet`` of rewrite rules, and applying these rules to a dask, or a single task. Matching is done syntactically only, and assumes all functions have fixed arity. Addresses #79.

Current capabilities:
- Bottom up rewriting of terms
- Lazy generation of matches from a matchset
- Nonlinear patterns (variables can be in a pattern more than once)

Demonstration:

```
from dask.rewrite import *
from dask.core import get

def f(a):
    return a

def g(a, b):
    return a + b

def h(a):
    return sum(a)

r = RuleSet(RewriteRule((g, (f, 'x'), 'y'), (f, (g, 'x', 'y')), ('x', 'y')),
            RewriteRule((g, (h, ['x']), 'x'), (g, 'x', 'x'), ('x',)))

dsk = {'x': (g, (f, 1), 2),
       'y': (g, (h, [1]), 1)}

new_dsk = r.rewrite(dsk)
```

The code was pulled and modified from some work I've been doing on matching for SymPy, and is thus mostly optimized for computer algebra applications. Even so, the above rewrite operation runs in ~260 microseconds on my 6 year old laptop.

Todo:
- [x] Add tests
- [x] Settle on interface design

### CODE REVIEW

**Comment by jcrist** [comment]
Possible optimizations:

- Make the `RuleSet` deterministic, removing any backtracking. This well be most beneficial if the patterns are small/few in number.
- Only apply to selected `task`s, rather than a whole `dask`.
- Make `subs` accept a dict, rather than iteratively calling it with `key, value` pairs.

**Comment by jcrist** [comment]
I just pushed a couple bug fixes/code cleanup. Still needs tests, but at this point I think it's as fast as it's going to get without the algorithm changing. Of the 3 possibilities I suggested above, number 1 has the biggest opportunity to cut speed. The downside is it makes adding rules after set creation more difficult (however, I'm not sure that's even necessary). Storing choice points for backtracking is one of the costlier operations that we can avoid.

I also thought of an easy way to add support for varargs functions. I haven't benchmarked it, but I'm fairly certain it will add minimal overhead.

Questions:

1. Is the current interface acceptable? User interacts with `RewriteRule` and `RuleSet`. Main function is `rewrite`.

2. Are the current capabilities (speed and features) sufficient for the task? If so, no need to spend more time improving it.

**Comment by mrocklin** [comment]
1.  The interface seems good to me
2.  I think so.  I need to spend some time and come up with a use case that stresses this

In general this looks awesome.  The main onus is probably now on me to figure out all the ways in which we can use it.  I wonder if [lazify_task](https://github.com/ContinuumIO/dask/blob/master/dask/bag/core.py#L29-L47) might be such a use case.

Regarding speed I'm in the process of benchmarking things.  From what I've seen so far 250us seems totally fine.

It would be reassuring to see tests.  Maybe also a doc page describing both how-to-use and, if you're interested, how-it-works.

**Comment by mrocklin** [comment]
Something like the example at the top would also be useful as a doctest/docstring

File: dask/rewrite.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        n = N.edges.get(VAR, None)
+        if n and restore_state_flag:
+            matches.pop()
+        if n and not restore_state_flag:
+            stack.append((S.copy(), N))
+            matches.append(S._term)
+            S.skip()
+            N = n
+        else:
+            n = N.edges.get(S.current, None)
```
Comment (mrocklin) [inline dask/rewrite.py]: What if this is not hashable?

**Comment by jcrist** [comment]
I finally found some time to work on tests and docs now. Question: currently the matcher finds the most general patterns first (i.e. `(f, 1, 1)` matches `(f, x, x)` before `(f, x, 1)`). I have a version that matches the most specific patterns first, which is what people probably would expect. However, this is slightly slower. The example up top runs in 300 microseconds instead. If that's acceptable, I think matching most specific first is better.

File: dask/rewrite.py (hunk 1 above)
Comment (jcrist) [inline dask/rewrite.py]: With this algorithm, there's no way around that. Nets are represented as dicts, so all terms must be hashable. I have a deterministic matcher that compiles down to python functions, which wouldn't depend on hashability. It's not quite done yet though, but I think it's a better plan in the end. It's *much* faster too.

**Comment by mrocklin** [comment]
At the moment the threaded scheduler has about a [1ms overhead](http://dask.pydata.org/en/latest/scheduling.html#performance).  The tasks on which I currently want to use this have multi-second durations.  Adding millisecond overhead is totally acceptable.  This might become less ok if we get sloppy and do this many times.  For now I say go for using the specific matchings match first.

Also, FWIW, I currently have very small sets of patterns (one).  I expect that this might grow to ten or tens, but not to thousands.

**Comment by mrocklin** [comment]
I'm using this to push down column selections to the underlying data structures

```python
a, b, c = '~a', '~b', '~c'
from dask.rewrite import RuleSet, RewriteRule

rewrite_rules = RuleSet(RewriteRule((getitem, (pframe.get_partition, a, b), c),
                                    (pframe.get_partition, a, b, c),
                                    (a, b, c)))

```

Currently failing because I have non-hashable DataFrame objects in the dask.  Any way to treat non-hashables as literals?

**Comment by jcrist** [comment]
Alright. I'll push the specific first version, finish the docs and tests. When the deterministic matcher is done, we can use it here instead. That'll remove the hashing requirement as well.

**Comment by jcrist** [comment]
Alright. I'll push the specific first version, finish the docs and tests. When the deterministic matcher is done, we can use it here instead. That'll remove the hashing requirement as well.

**Comment by jcrist** [comment]
I think I could make that work. Matching non-hashables to variables should work, what can't work is matching to patterns containing non-hashable objects. I'll see what I can do.

**Comment by mrocklin** [comment]
Hrm, yes, I see how that would be tricky.

**Comment by jcrist** [comment]
Got it working with non-hashables. Just needed a `try-except` block. Easy fix.

File: dask/rewrite.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        term: dask, or a task
+            The item to be rewritten
+        strategy: str, optional
+            The rewriting strategy to use. Options are "bottom_up" (default),
+            or "top_level".
+        """
+        func = strategies[strategy]
+        if isinstance(term, dict):
+            return dict((k, func(self, v)) for (k, v) in term.items())
+        return func(self, term)
```
Comment (mrocklin) [inline dask/rewrite.py]: I suggest that we keep RuleSets specific to just terms and use other mechanisms (like `toolz.valmap`) to rewrite on many terms within a dict.

**Comment by jcrist** [comment]
@mrocklin : I just pushed an update that matches the most specific pattern first, and allows for non-hashable terms in the subject (not the pattern). I tested it locally on some simple expressions, and it seems to work fine. Does it work for your problem?

I also improved the docstrings, and added tests. I don't have time right now to write up a docs page, I may have time to do that next week.

**Comment by mrocklin** [comment]
Is it possible that you haven't git added a test file to the repository?

**Comment by jcrist** [comment]
It is. Fixed.

**Comment by mrocklin** [comment]
OK, this is working really well.  Well, at least one one of my two problems.  Maybe I can work around it?  Not sure.

I think I'm running into the nested+unhashable problem.  The following test fails for me.

```python
from dask.rewrite import RuleSet, RewriteRule

a, b, c, d, e = '~a', '~b', '~c', '~d', '~e'  # vars
f, g, h = 'f', 'g', 'h'  # functions

r = RuleSet(RewriteRule((f, (g, a, b, c, None), d),
                        (g, a, b, c, d),
                        (a, b, c, d)))

def test_nested_and_unhashable():
    term = (h, (f, (g, 1, 2, dict(), None), 3))
    expected = (h, (g, 1, 2, dict(), 3))

    assert r.rewrite(term) == expected
```

**Comment by jcrist** [comment]
That test fails because none of the above are *tasks*. To differentiate between just tuples, and tasks, the algorithm only traverses into sub-arguments for tuples with the first element as a callable, or lists. Try it with actual functions for `f`, `g`, and `h`, it works fine then. I *could* make this traverse into any tuple, if that'd be needed.

**Comment by mrocklin** [comment]
Ah, ok.  Got it.

**Comment by mrocklin** [comment]
A bit curious on how separable this could be.  Am I right in assuming that we could replace `dask.istask` with `isinstance(..., tuple)` and still achieve desired behavior?  Other than that the only dependent functionality is `subs` which could arguably be duplicated (with `istask` again replaced).  

If so what are your thoughts on making `rewrite` a standalone library?  We could distribute it along with dask in the same way we do `pframe`.  (separate directory, shared setup.py file)

**Comment by jcrist** [comment]
Yeah, that's why it was so easy to get it working for dask; I didn't need to change much. Here's the interface that's needed to hook in:

1. A `Traverser` object (must be copyable), that holds the state for performing a preorder-traversal on a term. The traverser must provide access to the current term, and methods to proceed to the next term, or skip over all subterms.
2. `subs` for rewriting terms
3. `head` and `args` for getting the head of a term, and its arguments respectively

And that's it. So it could be easily separated. The bulk of the work is the `Traverser`, which basically defines the structure of the terms. Would separating it be something you'd want to do?

**Comment by mrocklin** [comment]
Mostly I wouldn't mind exposing this functionality to projects that aren't dask but still fundamentally manipulate nested tuple expressions (which I'd like to see gain traction.)  If we generalize `head`, `args` and `subs` slightly it seems like we lose the dask dependency.  I can give this a shot if you're feeling done.  Will merge this if so.

**Comment by jcrist** [comment]
Nah, I can take it, shouldn't take more than an hour or two.

If it's generalized, users need to define `head`, `args` and `subs`, and everything else will *just work*. My thought would be to pass these in to the creation of the `RuleSet`. Should a curried version be made specific for dask that includes the dask specific functions already?

**Comment by mrocklin** [comment]
Ah, in my mind I had "generalized" this to the specific case of nested tuples (without using the dask.istask and dask.subs methods).  Or maybe what I want is actually the existing implementation but slightly external to dask.  

What is the performance hit to full generalization?  

Perhaps fully generalize with nested tuples or tasks as sane default?

**Comment by mrocklin** [comment]
Or we can just merge and work on other things, generalizing later if it becomes useful.

**Comment by mrocklin** [comment]
Actually, I'm now pro "merge-now, generalize later" if you're cool with that.

**Comment by jcrist** [comment]
Sounds good to me. If we want to generalize later, it shouldn't be hard.

**Pull request merged**